File: day10/main.py
import functools

with open("input") as f:
    arr = [0] + [ int(x) for x in f.read().split() ]

arr.sort()

last_index = len(arr) - 1
    
@functools.lru_cache(None)
def gold(i):
    if i + 2 == last_index:
        if arr[-1] - arr[i] <= 3:
            return 2
        return 1
    if i + 1 == last_index:
        return 1
    if i == last_index:
        return 1
    r = gold(i + 1)
    if arr[i + 2] - arr[i] <= 3:
        r += gold(i+2)
        if arr[i + 3] - arr[i] <= 3:
            r += gold(i+3)
    return r

print("Part two:", gold(0))

File: day10/main.hs
import           Data.List

main :: IO ()
main = do
  ns <- map read . lines <$> readFile "input"
  putStrLn $ "Part one: " ++ show (silver ns)

silver :: [Integer] -> Integer
silver js = oneJoltDiff * threeJoltsDiff
  where (_, oneJoltDiff, threeJoltsDiff) =
          foldl go (0, 0, 1) $ sort js
        go (currentJolt, oneDiffCount, threeDiffCount) j =
          case j - currentJolt of
            1 -> (j, oneDiffCount + 1, threeDiffCount)
            3 -> (j, oneDiffCount, threeDiffCount + 1)
            _ -> (j, oneDiffCount, threeDiffCount)
